diesel/src/query_builder/ast_pass.rs

```diff
@@ -69,7 +69,10 @@ where
         }
     }
 
-    pub(crate) fn debug_binds(formatter: &'a mut Vec<&'b dyn fmt::Debug>, backend: &'b DB) -> Self {
+    pub(crate) fn debug_binds(
+        formatter: &'a mut Vec<Box<dyn fmt::Debug + 'b>>,
```

Comment:
Whats the benefit of using `Box<>` here?

Reply:
That's required because we cannot store references there anymore due to the fact that it is not possible to (up)cast a trait object (something that's not sized) to another trait object (dyn Debug). That would be required for the changed signature of `push_bind_param`

Reply:
Ah I should have commented this yesterday... This probably means that accepting this change means we won't ever be able to use &dyn AnyHelperTraitForBindingPush anywhere in the `push_bind_param` interface except for debugging, because the impact on perf would be too large.

I suspect we probably wouldn't want to do that in most cases because likely that would prevent the compiler form const-optimizing it, but maybe for `push_bind_param_value_only` that would end up being acceptable... so I'm not sure if that isn't risky for the future...
I was curious as well to see what the concrete use-case that would be provided as answer to https://github.com/diesel-rs/diesel/pull/3548#issuecomment-1468680101 would look like to try and make an informed weighing of what it enables vs what it blocks.

Reply:
@Ten0 I'm not 100% sure that I can follow your argumentation there. Can you provide a few more details about your ideas around `AnyHelperTraitForBindingPush`?

Reply:
The idea is that since this prevents from using &dyn XX internally inside this function, this prevents any design where we would want to go with &dyn XXX at some point in some of the passes. Currently there is only `Debug` where that was the case, but I'm not sure if there's really no chance that we would want to do a design that uses this at some point, and I'm not sure if the gain of not having to add an extra `&` on the user side is worth it (if that part of the code did compile, otherwise it's definitely worth it).
That being said I don't have a strong opinion on it, just wanted to raise the point.

Reply:
Unfortunately just adding a `&` wasn't sufficient, because then we had something like `&'temp &'b str` which fixed the sizedness but got us a lifetime error because the reference now lives for `'temp` instead of `'b` as required by `fn push_bind_param`.

Reply:
OK that was my question at https://github.com/diesel-rs/diesel/pull/3548#discussion_r1138872331
LGTM then